### ISSUE TITLE
Inconsistency between binary search for kv chunk size and the actual splitting strategy. 

### ISSUE THREAD

**Issue by timzsu** [flashinfer-ai/flashinfer#726]
Hi, 

Thanks for the impressive work! I would like to report a bug related to attention scheduling and discuss possible solutions. 

# Bug description: 

I encountered a violation of the check `new_batch_size <= padded_batch_size, "new batch size should not exceed padded batch size"` in `scheduler.cuh` when using vllm with Cuda graph and flash infer backend. The root cause is an inconsistency between the treatment to zero-length kv (e.g. padded inputs) during binary search for appropriate kv chunk size and tile splitting. Namely, in [PrefillBinarySearchKVChunkSize](https://github.com/flashinfer-ai/flashinfer/blob/13de89657ab293ef95a7fbae516eaaa96a65a059/include/flashinfer/attention/scheduler.cuh#L85) when `kv_len_arr[i]=0` it doesn't contribute to `new_batch_size`, but in real splitting they do contribute due to [this line](https://github.com/flashinfer-ai/flashinfer/blob/13de89657ab293ef95a7fbae516eaaa96a65a059/include/flashinfer/attention/scheduler.cuh#L510). 

# Minimal working example: 

This is a handcrafted example only to show the problem. The parameters might not be realistic but it reproduces the problem I encountered with vllm flash infer. 

```python
import torch
import flashinfer

batch_size = 24
qo_len = 5
num_qo_heads = 40
head_dim = 128
kv_len = 5
num_kv_heads = 8
page_size = 16
total_num_pages = 10743

q = torch.randn(batch_size * qo_len, num_qo_heads, head_dim).to(0).half()
kv_shape = [total_num_pages, 2, page_size, num_kv_heads, head_dim]
kv_data_fp32 = torch.randn(*kv_shape, dtype=torch.float32).to(0)
kv_data = kv_data_fp32.half()

q_indptr_buffer = torch.empty(batch_size + 1).int().to(0)
kv_indptr_buffer = torch.empty(batch_size + 1).int().to(0)
kv_indices_buffer = torch.empty(total_num_pages).int().to(0)
kv_last_page_len_buffer = torch.empty(batch_size).int().to(0)
workspace_buffer = torch.empty(128 * 1024 * 1024, dtype=torch.int8).to(0)
wrapper = flashinfer.prefill.BatchPrefillWithPagedKVCacheWrapper(
    workspace_buffer,
    "NHD",
    use_cuda_graph=True,
    qo_indptr_buf=q_indptr_buffer,
    paged_kv_indptr_buf=kv_indptr_buffer,
    paged_kv_indices_buf=kv_indices_buffer,
    paged_kv_last_page_len_buf=kv_last_page_len_buffer,
    backend="fa2"
)
q_indptr_warmup = torch.arange(0, batch_size + 1).int() * qo_len
kv_indptr_warmup = torch.arange(0, batch_size + 1).int()
kv_indices_warmup = torch.arange(0, batch_size).int()
kv_last_page_len_warmup = torch.full(
    (batch_size,), page_size, dtype=torch.int32
)

wrapper.plan(
    q_indptr_warmup,
    kv_indptr_warmup,
    kv_indices_warmup,
    kv_last_page_len_warmup,
    num_qo_heads,
    num_kv_heads,
    head_dim,
    page_size,
    causal=True,
)

# warmup
s = torch.cuda.Stream()
s.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s):
    for _ in range(3):
        o = wrapper.run(q, kv_data)
torch.cuda.current_stream().wait_stream(s)
# capture
g = torch.cuda.CUDAGraph()
with torch.cuda.graph(g):
    o = wrapper.run(q, kv_data)

qo_indptr = torch.arange(0, batch_size + 1).to(0).int() * qo_len
kv_len_arr = torch.empty(batch_size + 1)
kv_len_arr[:12] = 512
kv_len_arr[12:21] = 511
kv_len_arr[21:] = 0
kv_indptr = torch.cumsum(kv_len_arr, 0).to(0).int()
kv_indices = torch.arange(0, kv_indptr[-1]).to(0).int()
kv_last_page_len = torch.full(
    (batch_size,), (kv_len - 1) % page_size + 1, dtype=torch.int32
).to(0)

wrapper.plan(
    qo_indptr,
    kv_indptr,
    kv_indices,
    kv_last_page_len,
    num_qo_heads,
    num_kv_heads,
    head_dim,
    page_size,
    q_data_type=torch.float16,
    kv_data_type=torch.float16,
)
```

# Possible solution

To solve this problem, we should make the treatment consistent. However, I am not sure whether it is preferred to assign a chunk to zero-length kv. Currently I assume the version in real splitting is preferred (i.e. assign a chunk to zero-length kv), so I suggest the following changes to the function [PrefillBinarySearchKVChunkSize](https://github.com/flashinfer-ai/flashinfer/blob/13de89657ab293ef95a7fbae516eaaa96a65a059/include/flashinfer/attention/scheduler.cuh#L85). After making this changes the error will disappear. 

```python
inline auto PrefillBinarySearchKVChunkSize(const bool enable_cuda_graph,
                                           const uint32_t max_batch_size_if_split,
                                           const std::vector<int64_t>& packed_qo_len_arr,
                                           const std::vector<int64_t>& kv_len_arr,
                                           const uint32_t qo_chunk_size,
                                           const uint32_t min_kv_chunk_size = 1) {
  const int64_t batch_size = packed_qo_len_arr.size();
  int64_t max_kv_len = 1;
  for (const int64_t& kv_len : kv_len_arr) {
    max_kv_len = std::max(max_kv_len, kv_len);
  }

  int64_t low = min_kv_chunk_size;
  int64_t high = max_kv_len;
  constexpr int64_t min_kv_len = 1;
  while (low < high) {
    const int64_t mid = (low + high) / 2;
    int64_t new_batch_size = 0;
    for (uint32_t i = 0; i < batch_size; ++i) {
      new_batch_size +=
          ceil_div(packed_qo_len_arr[i], qo_chunk_size) * ceil_div(std::max(kv_len_arr[i], min_kv_len), mid);
    }
    if (new_batch_size > max_batch_size_if_split) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }
  return std::make_tuple(enable_cuda_graph || low < max_kv_len, low);
}
```

BR

**Comment by timzsu** [comment]
Sorry, I mistakenly closed this issue. Feel free to lmk if you want me to create a PR :)

**Comment by yzh119** [comment]
Hi @timzsu  thanks for bringing this up, we are about to replace the scheduler with the v2 scheduler (mentioned in the section 3.3 of [the paper](https://arxiv.org/pdf/2501.01005)), eta is this weekend. So the binary search strategy will be forsaken.

But yes I'll definitely welcome a bugfix PR from you so that we can release a v0.2.0.post2 version that fixes this bug!

**Comment by timzsu** [comment]
I created a PR. Feel free to let me know if there is any problem!
